lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -219,20 +327,65 @@ static LogicalResult extractConcatToConcatExtract(ExtractOp op,
   return success();
 }
 
-LogicalResult ExtractOp::canonicalize(ExtractOp op, PatternRewriter &rewriter) {
-  // Narrow Mux
-  // extract(mux(c,a,b)) -> mux(c,extract(a),extract(b))
-  if (auto mux = dyn_cast_or_null<MuxOp>(op.input().getDefiningOp())) {
-    if (mux->hasOneUse()) {
-      auto newT = rewriter.createOrFold<ExtractOp>(
-          mux.getLoc(), op.getType(), mux.trueValue(), op.lowBit());
-      auto newF = rewriter.createOrFold<ExtractOp>(
-          mux.getLoc(), op.getType(), mux.falseValue(), op.lowBit());
-      rewriter.replaceOpWithNewOp<MuxOp>(op, mux.cond(), newT, newF);
-      return success();
-    }
+// Pattern matches on extract(f(a, b)), and transforms f(extract(a), extract(b))
+// for some known f. This is performed by analyzing all usage sites of the
+// result of f(a, b). When this transformation returns true, it mutates all the
+// usage sites of outerExtractOp to reference the newly created narrowed
+// operation.
+static bool narrowExtractWidth(ExtractOp outerExtractOp,
+                               PatternRewriter rewriter) {
+  auto *innerArg = outerExtractOp.input().getDefiningOp();
+
+  if (!innerArg) {
```

Comment:
I think this needs to check that `innerArg->hasOneUse()` because this will otherwise result in duplicating the computation (in the narrower width).

Reply:
`narrowOperationWidth` was written such that it rewrite the `innerArg` directly rather than the extract, and mutates the use sites to reference the mutated `innerArg`.

[This test](https://github.com/llvm/circt/pull/1424/files#diff-838d92f61a1cefe7d7b777fdd77339eef086cf28c27b61e568aefc344d189a53R340) validates that we won't do duplicate work.

Reply:
I see, this is handled by getLowestBitAndHighestBitRequired.  Cool, thanks.